fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/ExpressionStatisticCalculator.java

```diff
@@ -156,121 +156,306 @@ public ColumnStatistic visitCall(CallOperator call, Void context) {
+                    distinctValue);
         }
 
         private ColumnStatistic binaryExpressionCalculate(CallOperator callOperator, ColumnStatistic left,
                                                           ColumnStatistic right) {
-            double distinctValues = Math.max(left.getDistinctValuesCount(), right.getDistinctValuesCount());
+            final double minValue;
+            final double maxValue;
             double nullsFraction = 1 - ((1 - left.getNullsFraction()) * (1 - right.getNullsFraction()));
+            double distinctValues = Math.max(left.getDistinctValuesCount(), right.getDistinctValuesCount());
+            double averageRowSize = callOperator.getType().getTypeSize();
             switch (callOperator.getFnName().toLowerCase()) {
                 case FunctionSet.ADD:
-                    return new ColumnStatistic(left.getMinValue() + right.getMinValue(),
-                            left.getMaxValue() + right.getMaxValue(), nullsFraction,
-                            callOperator.getType().getTypeSize(),
-                            distinctValues);
+                case FunctionSet.MICROSECONDS_ADD:
+                case FunctionSet.DATE_ADD:
+                    minValue = left.getMinValue() + right.getMinValue();
+                    maxValue = left.getMaxValue() + right.getMaxValue();
+                    break;
                 case FunctionSet.SUBTRACT:
-                    return new ColumnStatistic(left.getMinValue() - right.getMaxValue(),
-                            left.getMaxValue() - right.getMinValue(), nullsFraction,
-                            callOperator.getType().getTypeSize(),
-                            distinctValues);
+                case FunctionSet.SECONDS_DIFF:
+                case FunctionSet.MICROSECONDS_SUB:
+                case FunctionSet.DATE_DIFF:
+                case FunctionSet.TIMEDIFF:
+                case FunctionSet.DATE_SUB:
+                    minValue = left.getMinValue() - right.getMaxValue();
+                    maxValue = left.getMaxValue() - right.getMinValue();
+                    break;
+                case FunctionSet.YEARS_DIFF:
+                    minValue = left.getMinValue() - right.getMaxValue();
+                    maxValue = left.getMaxValue() - right.getMinValue();
+                    distinctValues = distinctOfInterval(left.getMinValue() - right.getMaxValue(),
+                            left.getMaxValue() - right.getMinValue(), 3600 * 24 * 365);
+                    break;
+                case FunctionSet.MONTHS_DIFF:
+                    minValue = left.getMinValue() - right.getMaxValue();
+                    maxValue = left.getMaxValue() - right.getMinValue();
+                    distinctValues = distinctOfInterval(left.getMinValue() - right.getMaxValue(),
+                            left.getMaxValue() - right.getMinValue(), 3600 * 24 * 31);
+                    break;
+                case FunctionSet.WEEKS_DIFF:
```

Comment:
could we use a visitor class here? then we needn't a lot of switch case.

Reply:
Seems not worthy here, because the number of function is large, and it may be expended later, and the number operation is fixed, just 3 of them, which is `nullaryExpressionCalculate`、`unaryExpressionCalculate`、`binaryExpressionCalculate`、`multiaryExpressionCalculate`